src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+      this.layouts_.splice(index, 1);
+    }
+  }
+
+  /**
+   * Calculates the element's intersection with the viewport's rect.
+   *
+   * @param {!Element} element A regular or AMP Element
+   * @param {!LayoutRectDef} viewportRect
+   */
+  calcIntersectionWithViewport(element) {
+    return this.calcIntersectionWithParent(element,
+        getScrollingElement(this.document_));
+  }
+
+  /**
+   * Calculates the element's intersection with the parent's rect.
+   * You may optionally expand the parent's rectangle.
+   *
+   * @param {!Element} element A regular or AMP Element
+   * @param {!Element} parent
+   * @param {{dw: number=, dh: number=}=} opt_expand
+   * @param {number=} opt_dh
+   * @return {!LayoutRectDef}
+   */
+  calcIntersectionWithParent(element, parent, opt_expand) {
+    const {left, top} = LayoutElement.getScrolledPosition(element, parent);
+    const size = LayoutElement.getSize(element);
+
+    let {width: parentWidth, height: parentHeight} =
+        LayoutElement.getSize(parent);
+    if (opt_expand) {
+      parentWidth *= 1 + (opt_expand.dw || 0) * 2;
+      parentHeight *= 1 + (opt_expand.dh || 0) * 2;
+    }
+
+    return layoutRectLtwh(
+        left,
+        top,
+        Math.max(Math.min(left + size.width, parentWidth) - left, 0),
+        Math.max(Math.min(top + size.height, parentHeight) - top, 0)
+    );
+  }
+
+  getScrolledPosition(element, opt_parent) {
+    const layout = this.add(element);
+    return layout.getScrolledPosition(opt_parent);
+  }
+
+  getOffsetPosition(element, opt_parent) {
+    const layout = this.add(element);
+    return layout.getOffsetPosition(opt_parent);
+  }
+
+  getSize(element) {
+    const layout = this.add(element);
+    return layout.getSize(element);
+  }
+
+  /**
+   * Remeasures the element, and any other elements who's cached rects may have
+   * been altered by this element's mutation.
+   *
+   * This is meant to be called after any element mutation happens (eg.
+   * #buildCallback, #layoutCallback, viewport changes size).
+   *
+   * @param {!Element} element A regular or AMP Element
+   */
+  remeasure(element) {
+    const layer = LayoutElement.getParentLayer(element);
+    if (layer) {
+      layer.remeasure();
+    }
+  }
+
+  /**
+   * Changes the element's rect size.
+   *
+   * @param {!AmpElement} element An AMP Element
+   * @param {!SizeDef} size
+   * @param {boolean=} force Whether to skip approval/denial logic
+   * TODO
+   */
+  // changeSize(element, size, force = false) {
+  // }
+
+  /**
+   * Eagerly creates a Layer for the element.
+   * @param {!Element} element
+   */
+  declareLayer(element) {
+    this.declareLayer_(element);
+  }
+
+  /**
+   * Eagerly creates a Layer for the element.
+   * @param {!Element} element
+   * @return {!LayoutElement}
+   */
+  declareLayer_(element) {
+    const layout = this.add(element);
+    layout.declareLayer();
+    return layout;
+  }
+
+  onResize_() {
+    const layouts = this.layouts_;
```

Comment:
Should layers be able to resize individually? Or will that just be part of `remeasure()` call?

Reply:
Ditto: drive from viewport?